File: app/src/main/java/com/wyattconrad/cs_360weighttracker/model/Weight.java
package com.wyattconrad.cs_360weighttracker.model;

import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.ForeignKey;
import androidx.room.PrimaryKey;

import java.time.LocalDateTime;
import java.time.ZoneId;

@Entity(foreignKeys = @ForeignKey(entity = User.class, parentColumns = "id", childColumns = "user_id", onDelete = ForeignKey.CASCADE))
public class Weight {

    // Create the fields for the database
    // Primary key and ID for the weight
    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "id")
    private long mId;

    // Weight value
    @ColumnInfo(name = "weight")
    private double mWeight;

    // Date and time logged
    @ColumnInfo(name = "date_time_logged")
    private long mDateTimeLogged;

    // User ID that logged the weight, foreign key from user table
    @ColumnInfo(name = "user_id")
    private long mUserId;

    // Constructor for creating a new weight
    public Weight(double weight, long mUserId) {
        mWeight = weight;
        this.mUserId = mUserId;
        mDateTimeLogged = System.currentTimeMillis();
    }

    public Weight(double weight, long mUserId, LocalDateTime datetimelogged) {
        mWeight = weight;
        this.mUserId = mUserId;
        // Convert LocalDateTime to long
        ZoneId zoneId = ZoneId.systemDefault();
        mDateTimeLogged = datetimelogged.atZone(zoneId).toInstant().toEpochMilli();
    }

    // Getters and Setters
    // Get the id
    public long getId() {
        return mId;
    }
    // Set the id
    public void setId(long id) {
        mId = id;
    }

    // Get the weight
    public double getWeight() {
        return mWeight;
    }
    // Set the weight
    public void setWeight(double weight) {
        mWeight = weight;
    }

    // Get the date and time logged
    public long getDateTimeLogged() {
        return mDateTimeLogged;
    }
    // Set the date and time logged
    public void setDateTimeLogged(long datetimelogged) {
        mDateTimeLogged = datetimelogged;
    }

    // Get the user id
    public long getUserId() {
        return mUserId;
    }
    // Set the user id
    public void setUserId(long userid) {
        mUserId = userid;
    }

}

File: app/src/main/java/com/wyattconrad/cs_360weighttracker/model/Weight.kt
/*
 * Copyright (C) 2025 Wyatt Conrad. All rights reserved.
 *
 * This file is part of the CS-360 Weight Tracker project.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.wyattconrad.cs_360weighttracker.model

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.Index
import androidx.room.PrimaryKey
import java.time.LocalDateTime

/**
 * The weight entity
 * @author Wyatt Conrad
 * @version 1.0
 */
// Set a foreign key to the user table
@Entity(
    tableName = "weights",
    foreignKeys = [ForeignKey(entity = User::class, parentColumns = ["id"], childColumns = ["user_id"], onDelete = ForeignKey.CASCADE)],
    indices = [Index(value = ["user_id"])]
)
data class Weight (

    // Create the id as the primary key
    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "id")
    var id: Long = 0L,

    // Create the weight column
    @ColumnInfo(name = "weight")
    var weight: Double,

    // Create the date time logged column
    @ColumnInfo(name = "date_time_logged")
    val dateTimeLogged: LocalDateTime,

    // Create the user id column
    @ColumnInfo(name = "user_id")
    val userId: Long,
) {
    /**
     * Constructor for the weight entity
     * @param weight The weight
     * @param userId The user id
     */
    constructor(weight: Double, userId: Long) : this(
        weight = weight,
        userId = userId,
        dateTimeLogged = LocalDateTime.now()
    )
}
